src/main/kotlin/RBSearchTree.kt

```diff
@@ -0,0 +1,276 @@
+
+class RedBlackTreeNode<K, V> (key: K,value: V): TreeNode<K, V, RedBlackTreeNode<K, V>> (key, value) {
+    var parent: RedBlackTreeNode<K, V>? = null
+    var isRed: Boolean = false
+}
+
+class RedBlackTreeSearch<K : Comparable<K>, V>: TreeSearch<K, V, RedBlackTreeNode<K, V>>() {
+    private fun rightRotate(node: RedBlackTreeNode<K, V>) {
+        val temp: RedBlackTreeNode<K, V>? = node.leftChild //temp - old left child node
+        /** instead of the left child of the node
+         * we put the right child of the temp */
+        node.leftChild = temp?.rightChild
+        temp?.rightChild?.parent = node
+        /** replace the node of the temp,
+         * keeping its right child */
+        temp?.parent = node.parent
+        if (node.parent == null) {
+            root = temp
+        } else {
+            if (node.parent?.rightChild == node) {
+                node.parent?.rightChild = temp
+            } else {
+                node.parent?.leftChild = temp
+            }
+        }
+        //set the node as the right child of the temp
+        temp?.rightChild = node
+        node.parent = temp
+    }
+
+    private fun leftRotate(node: RedBlackTreeNode<K, V>) {
+        val temp: RedBlackTreeNode<K, V>? = node.rightChild //temp - old right child node
+        /** instead of the right child of the node
+         * we put the left child of the temp */
+        node.rightChild = temp?.leftChild
+        temp?.leftChild?.parent = node
+        /** replace the node of the temp,
+         * keeping its left child */
+        temp?.parent = node.parent
+        if (node.parent == null) {
+            root = temp
+        } else {
+            if (node == node.parent?.leftChild) {
+                node.parent?.leftChild = temp
+            } else {
+                node.parent?.rightChild = temp
+            }
+        }
+        //set the node as the left child of the temp
+        temp?.leftChild = node
+        node.parent = temp
+    }
+
+    private fun balanceRemove(node: RedBlackTreeNode<K, V>) {
+        val parent = node.parent ?: return
+        val bro = if (parent.leftChild == node) parent.rightChild else parent.leftChild
+        /** always, when we call balancing, the node has a brother
+         * that is not equal to null (otherwise the tree would not be balanced) */
+        if (bro!!.isRed) { //if brother is red, change his color

```

Comment:
Возможно для таких случаев `bro!!.isRed` лучше сделать доп функцию вида:
```kotlin
fun isRedNode(node: RedBlackTreeNode<K, V>?) {
    if (node == null) return false
    return node.isRed
}
```


Или добавить более валидную с точки зрения пользователя проверку на исключительную ситуацию, когда ошибка выбрасывается более интересная чем NullPointer
```suggestion
if (bro == null) throw RuntimeException("Tree invalid balanced")
if (bro.isRed) {
   // Ваш код
}
```